Java: Query to detect hash without salt

luchua-bc opened the pull request:
In cryptography, "salt" is random data that are used as an additional input to a one-way function that hashes a password or pass-phrase. It makes dictionary attacks more difficult.

Without a salt, it is much easier for attackers to pre-compute the hash value using dictionary attack techniques such as rainbow tables to crack passwords.

This type of vulnerabilities is categorized as CWE-759 "Use of a One-Way Hash without a Salt". This query detects issues of this type.

Please consider to merge the PR. Thanks.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.qhelp:
```diff
@@ -0,0 +1,29 @@
+<qhelp>
+
+  <overview>
+    <p>In cryptography, "salt" is random data that are used as an additional input to a one-way function that hashes a password or pass-phrase. It makes dictionary attacks more difficult.</p>
```
```suggestion
    <p>In cryptography, a salt is some random data used as an additional input to a one-way function that hashes a password or pass-phrase. It makes dictionary attacks more difficult.</p>
```

luchua-bc replied:
Thanks @smowton for helping to review this PR. I've made the change.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,93 @@
+/**
+ * @id java/hash-without-salt
+ * @name Use of a One-Way Hash without a Salt
```
```suggestion
 * @name Use of a hash function without a salt
```

smowton replied:
(because all hash functions are one-way)

luchua-bc replied:
Changed

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  /** Holds for additional steps such as `passwordStr.getBytes()` */
+  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
```
This is very general, propagating all method arguments to all results and qualifiers -- suggest restricting this to particular cases you're interested in.

luchua-bc replied:
Added the type constraint

smowton replied:
So it's conducting taint from the arg of a digest or update to the hash accumulator? Why do we want that in the context of tracking taint from a password to a digest call? Could you add a test that only works with the additional-taint-step implemented, or indicate one that requires it now?

luchua-bc replied:
Without the additional taint step, somehow no correct result will be returned. So there must be something wrong with the logic. I will further investigate.

luchua-bc replied:
Sorry I confused myself. I removed the specific method names and just specified the object type. 

This additional taint step is for the check of:
```
// invoke `md.digest()` with only one call of `md.update(password)`, that is, without the call of `md.update(digest)`
```

Without it, the following invocation won't be detected since it's a separate method access:
```
	// BAD - Hash without a salt.
	public void getSHA256Hash2(String password) throws NoSuchAlgorithmException {
		MessageDigest md = MessageDigest.getInstance("SHA-256");
		md.update(password.getBytes());
		byte[] messageDigest = md.digest();
	}
```

And the part <code>pred.asExpr() = ma.getAnArgument()</code> also helps since the sink below is the method access itself instead of its argument:
```
    // invoke `md.digest(password)` without another call of `md.update(salt)`
    exists(MethodAccess mda |
      sink.asExpr() = mda and
```

Please let me know if we can further improve it. Thanks.

smowton replied:
Isn't that only the case because you define `sink.asExpr() = mda.getQualifier()`? What if the sink was the argument to the `update` call, then the rest of the definition of an interesting update call (looking for sibling digest() and other update calls) would do the rest of the work?

luchua-bc replied:
Thanks a lot. That helps to remove the additional taint step.

luchua-bc commented:
Added <code> isSanitizer</code> check to detect the scenario of passwords concatenated with a salt to reduce the number of false positives.

smowton commented:
Let me know when you regard this as ready for evaluation

luchua-bc commented:
Thanks for the follow-up, I think it’s ready for evaluation now. Please proceed with moving forward.

JarLob commented:
The query produced a lot of false positives. Some cases I have noticed that maybe could be improved:
1. If there is a hash wrapper that has `MessageDigest` as a member like https://github.com/is/jsch/blob/addb8e3a0ebd71cbbf51cf6ba746e8e128df713c/src/main/java/com/jcraft/jsch/jce/SHA256.java it warns about the usage of `update` function even if the wrapper function is called multiple times. I think the real sink should be `digest` and the warning shown only if there is a proof the update was called only once with a password.
2. The query also picks cases when the password is concatenated with a salt as a string first. For example:
```
String salt = getSalt()
hash(a.password + salt);
...
String hash(String payload) {
  MessageDigest alg = MessageDigest.getInstance("MD5");
  return alg.digest(payload.getBytes(StandardCharsets.UTF_8));
}
```
3. The regex `(?i).*pass(wd|word|code|phrase).*` also picks up `passwordHash`. Sometimes a hash is used as an input for a hashing function. Should be easy fix to exclude sources that have `hash` already in the name.

luchua-bc commented:
@JarLob I've modified the query to cover the second and the third scenarios. 

For the first one, I added a test case with the provided `SHA256` Java class. However, the query didn't report false positives you mentioned. I think the reason is that it's just a library class to be invoked by other programs. Would you please provide more details on how to reproduce the issue? Thanks.

JarLob commented:
What I meant it is giving false positives about the usage of `SHA256` in https://github.com/is/jsch project. Try running the query on the repository.

luchua-bc commented:
I've modified the query to check wrapper calls. Now the query doesn't report any FPs with the https://github.com/is/jsch project. However, it won't detect any call through wrapper classes/methods either as shown in the test cases.

As checking method invocation stacks is pretty common with CodeQL, I'm wondering whether there is a better way  to model call graph navigation for this query. The goal is to detect a `md.update()` call that's different from the call for handling the password, which is either `md.digest(password)` or `md.update(password)`. Could you shed some light on this? Thanks in advance for your help.

luchua-bc commented:
I've implemented a solution, which seems to fully work with the sample project and wrapper classes as shown in the test cases. Call graph navigation is properly validated when a method call is against a concrete class type whose implementation can be checked. Please review. Thanks.

smowton commented:
Suggested improvement to the architecture here, using three configurations:

Configuration 1: source = passwords, sink = md.update or md.digest calls. This tells us "is this update/digest integrating a password?". Some of the hard-coded code structure in the existing sanitizer could become a sanitizer for this configuration -- e.g. concatenating the password with something else is a sanitizer.

Configuration 2: source = MessageDigest constructors, sink = md.update / md.digest calls that match configuration 1 (i.e., which integrate a password). This tells us "does this MD5 instance at some point integrate a password?"

Configuration 3: source = MessageDigest constructors that match configuration 2, sink = md.digest calls (not update calls, i.e. the sink is the final digest), sanitizer = md.update or md.digest calls that *do not* match configuration 1 (i.e., which integrate something other than the password, perhaps a salt). This tells us "does this MD5 instance integrate a password but not a salt?"

This way as much as possible of the hard-coded code structure goes away, with passwords allowed to flow arbitrarily to update/digest calls and MessageDigest instances allowed to flow arbitrarily from their constructor through to the final digest(...) call, which should be considered the sink (so the user-facing message is "you allocate a MessageDigest *here*, it the final hash comes out *here*, and it appears to pick up a password *here* but not a salt")

JarLob commented:
There are still many false positives with the last query. See below. The suggestions should improve FP ratio.

- https://lgtm.com/projects/b/lucee/lucee/snapshot/e27abe6892683e8130fe5acd9ae057a86cdc7d8d/files/lucee-java/lucee-core/src/lucee/commons/digest/Hash.java?sort=name&dir=ASC&mode=heatmap#L152
- https://lgtm.com/projects/g/pingidentity/ldapsdk/snapshot/16004e6482c394857f1b4d687f37a71ba18b3c01/files/src/com/unboundid/ldap/listener/SaltedMessageDigestInMemoryPasswordEncoder.java?sort=name&dir=ASC&mode=heatmap#L256

luchua-bc commented:
Thanks for the suggested improvement. I will work on it and let you know if I have some questions.

luchua-bc commented:
Thanks @smowton for the detailed explanation. I tried a lot of times but still failed to implement it right. 

What I have is as follows:

Configuration 1:
```
/** Taint configuration tracking flow from an expression whose name suggests it holds password data to a method call that generates a hash without a salt. */
class PasswordHashConfiguration extends TaintTracking2::Configuration {
  PasswordHashConfiguration() { this = "PasswordHashConfiguration" }

  override predicate isSource(DataFlow::Node source) { source.asExpr() instanceof PasswordVarExpr }

  override predicate isSink(DataFlow::Node sink) {
    exists(
      MethodAccess ma // invoke `md.update(password)` without the call of `md.update(digest)`
    |
      sink.asExpr() = ma.getArgument(0) and
      (
        ma.getMethod() instanceof MDUpdateMethod or // md.update(password)
        ma.getMethod() instanceof MDDigestMethod // md.digest(password)
      )
    )
  }

  /**
   * Holds if a password is concatenated with a salt then hashed together through the call `System.arraycopy(password.getBytes(), ...)`, for example,
   *  `System.arraycopy(password.getBytes(), 0, allBytes, 0, password.getBytes().length);`
   *  `System.arraycopy(salt, 0, allBytes, password.getBytes().length, salt.length);`
   *  `byte[] messageDigest = md.digest(allBytes);`
   * Or the password is concatenated with a salt as a string.
   */
  override predicate isSanitizer(DataFlow::Node node) {
    exists(MethodAccess ma |
      ma.getMethod().getDeclaringType().hasQualifiedName("java.lang", "System") and
      ma.getMethod().hasName("arraycopy") and
      ma.getArgument(0) = node.asExpr()
    ) // System.arraycopy(password.getBytes(), ...)
    or
    exists(AddExpr e | node.asExpr() = e.getAnOperand()) // password+salt
  }
}
```

Configuration 2:
```
class PasswordDigestConfiguration extends TaintTracking::Configuration {
  PasswordDigestConfiguration() { this = "PasswordDigestConfiguration" }

  override predicate isSource(DataFlow2::Node source) {
    exists(MDConstructor mc | source.asExpr() = mc)
  }

  override predicate isSink(DataFlow2::Node sink) {
    exists(MethodAccess ma |
      (
        ma.getMethod() instanceof MDUpdateMethod or
        ma.getMethod() instanceof MDDigestMethod
      ) and
      exists(PasswordHashConfiguration cc | cc.hasFlowToExpr(ma.getAnArgument())) and
      sink.asExpr() = ma.getQualifier()
    )
  }
}
```

Configuration 3:
```
class HashWithoutSaltConfiguration extends TaintTracking::Configuration {
  HashWithoutSaltConfiguration() { this = "HashWithoutSaltConfiguration" }

  override predicate isSource(DataFlow::Node source) {
    exists(MDConstructor mc | source.asExpr() = mc)
  }

  override predicate isSink(DataFlow::Node sink) {
    exists(MethodAccess ma |
      ma.getMethod() instanceof MDDigestMethod and
      sink.asExpr() = ma.getQualifier()
    )
  }

  /** Holds if `md.update` or `md.digest` calls integrate something other than the password, perhaps a salt. */
  override predicate isSanitizer(DataFlow::Node node) {
    exists(MethodAccess ma |
      (
        ma.getMethod() instanceof MDUpdateMethod
        or
        ma.getMethod() instanceof MDDigestMethod and ma.getNumArgument() != 0
      ) and
      node.asExpr() = ma.getQualifier() and
      not exists(PasswordHashConfiguration cc | cc.hasFlowToExpr(ma.getAnArgument()))
    )
  }
}
```

I cannot use configuration 2 in configuration 3 as the source, or else CodeQL will complain circular reference. And source of configuration 2 and configuration 3 seems to be identical.

So I must not understand your suggestion entirely. Would you please provide more advice? Thanks in advance for your help.

smowton commented:
This is because dependent taint-tracking or data-flow configurations must use different instances of the taint-tracking library (TaintTracking, TaintTracking2, TaintTracking3 etc)

Independent configurations (ones that don't refer to each other) may use the same instance, though I'm told there are some small performance benefits to not doing so.

luchua-bc commented:
There are only TaintTracking and TaintTracking2 in the [taint-tracking library](https://github.com/github/codeql/tree/8d4f7e2db7ec3b7559a3e358fd1b8a7203ce7474/java/ql/src/semmle/code/java/dataflow). Shall I create my own TaintTracking3?

smowton commented:
Yes, please do -- the TaintTrackingImpl.qll files should be identical; the TaintTrackingParameter.qll files differ in the numbered version of DataFlow they import.

luchua-bc commented:
Thanks @smowton. I will add TaintTracking3. Do those three configurations look like what you want except the source of configuration 3, which shall be as follows?
```
  override predicate isSource(DataFlow::Node source) {
    exists(PasswordDigestConfiguration pc | pc.hasFlow(source, _))
  }
```

Thanks,
@luchua-bc

intrigus-lgtm commented:
@smowton shouldn't this new TaintTracking lib then be added to this list?
(Although I don't know how https://github.com/github/codeql/blob/main/config/sync-files.py would work)

https://github.com/github/codeql/blob/cd20163f6ed506e56f5627b7a61f8bb7157d077d/config/identical-files.json#L34-L50

smowton commented:
Yes, I suppose so -- and yes those configurations look right to me.

luchua-bc commented:
Hi @smowton,

I've committed the change to use three configurations to detect password hash without salt. It helps to eliminate false positives, now there is no FP reported for those three projects. 

However, the query doesn't detect wrapper method calls of password hash without salt anymore e.g. the following test case:
https://github.com/github/codeql/blob/ff1ed3a012e720da3a04779ee0c177dd2ba7b4c0/java/ql/test/experimental/query-tests/security/CWE-759/HashWithoutSalt.java#L95-L101

Sorry that I'm still not sure how to bridge them together. Could you shed some light on it?

Thanks a lot,
@luchua-bc

luchua-bc commented:
@smowton Would you please provide some suggestions when you have a chance? Thanks.

smowton commented:
OK, debugging strategy: replace the top-level query with
```
from DataFlow::PartialPathNode source, DataFlow::PartialPathNode sink, int distance, HashWithoutSaltConfiguration cc
where cc.hasPartialFlow(source, sink, distance)
select sink, source, sink, "$@ is hashed without a salt.", source, "The password"
```
and replace `import DataFlow::PathGraph` with `import DataFlow::PartialPathGraph`. Then we can see where the sources flow to.

Looks like all the flows out from the allocation in the `MessageDigest.getInstance` call in SHA256.java flow out of the `init()` call in `getSHA256Hash7`. `getSHA256Hash6` never calls `init()` so the flow never gets there.

Let's try altering the test to add an `init()`. That produces an extra path from SHA256.java:8 to SHA256.java:19, which is what we want -- we don't get an extra result though because the final selected flow-path is constructor-to-digest, not password-to-digest.

smowton commented:
Unfortunately, that reveals a major problem with my suggested approach here -- we're also flagging the final GOOD example getSHA256Hash7, because it integrates its salt at the same callsite in SHA256.java as a password is integrated, thus not matching the sanitizer intended to block flow where a salt is included.

It may be that my 3-config approach is losing too much context between the related paths here and so will also be very inaccurate. Back to the drawing board I think.

luchua-bc commented:
Thanks @smowton for the update. I'll also ponder how to improve this query and will try to draft an improved version then ask your advice.

luchua-bc commented:
Hi @smowton, I've modified the query to only detect the scenario without wrapper calls, which is the most common scenario in which we are interested and helps to reduce FPs. Now the query works with all three sample projects.

Please review. Thanks.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+/** Holds if `MethodAccess` is a single hashing call. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not exists(MethodAccess ma2 | hasParentCall(ma2, ma))
```
Explain what `not ... hasParentCall` is intended to achieve here?

luchua-bc replied:
I've added the explanation that the intention here is to only consider sinks of <code>MDHashMethod</code> that are not invoked by a wrapper method, which could invoke MDHashMethod in another call stack, to reduce FPs.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+}
+
+/** Taint configuration tracking flow from an expression whose name suggests it holds password data to a method call that generates a hash without a salt. */
+class HashWithoutSaltConfiguration extends TaintTracking::Configuration {
```
Since we now only have one Configuration I guess all DataFlow2,3,... can be discarded from this PR

luchua-bc replied:
Removed <code>DataFlow2</code> and the newly added <code>TaintTracking3</code> since they are no longer being used.

luchua-bc commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+  }
+}
+
+/** Holds if `Expr` e is an operand of `AddExpr`. */
```
Thanks @smowton for reviewing the PR again. I've made the change.

luchua-bc commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+/** Holds if `Expr` e is an operand of `AddExpr`. */
+predicate hasAddExpr(AddExpr ae, Expr e) {
+  ae.getAnOperand() = e or
+  hasAddExpr(ae.getAnOperand(), e)
```
Updated

luchua-bc commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
+predicate hasAnotherHashCall(MethodAccess ma) {
+  exists(MethodAccess ma2, DataFlow2::Node node1, DataFlow2::Node node2 |
+    ma2 instanceof MDHashMethodAccess and
```
Updated

luchua-bc commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+    ma instanceof MDHashMethodAccess and
+    not hasAnotherHashCall(ma)
+  )
+}
```
Updated

luchua-bc commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,164 @@
+/** Holds if `MethodAccess` ma is invoked by `MethodAccess` ma2 either directly or indirectly. */
+predicate hasParentCall(MethodAccess ma2, MethodAccess ma) {
+  ma.getCaller() = ma2.getMethod() and
+  not ma2 instanceof MDHashMethodAccess
```
Changed

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+  MessageDigest() { this.hasQualifiedName("java.security", "MessageDigest") }
+}
+
+class MDConstructor extends StaticMethodAccess {
```
Add docstring or make private

luchua-bc replied:
Added docstring

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends Expr {
+  PasswordVarExpr() {
+    this.(VarAccess).getVariable().getName().toLowerCase().regexpMatch(getPasswordRegex()) and
+    not this.(VarAccess).getVariable().getName().toLowerCase().matches("%hash%") // Exclude variable names such as `passwordHash` since their values were already hashed
```
```suggestion
class PasswordVarAccess extends VarAccess {
  PasswordVarAccess() {
    exists(string name | name = this.getVariable().getName().toLowerCase() |
      name.regexpMatch(getPasswordRegex()) and not name.matches("%hash%")
    )
```

luchua-bc replied:
Updated

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+}
+
+/** Holds if `Expr` e is a direct or indirect operand of `ae`. */
+predicate hasAddExpr(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
```
```suggestion
predicate hasAddExprAncestor(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
```

luchua-bc replied:
Updated

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+predicate hasAddExpr(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
+
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
+predicate hasAnotherHashCall(MethodAccess ma) {
```
```suggestion
predicate hasAnotherHashCall(MDHashMethodAccess ma) {
```

luchua-bc replied:
Updated

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+/** Holds if `Expr` e is a direct or indirect operand of `ae`. */
+predicate hasAddExpr(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
+
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
```
```suggestion
/** Holds if `MDHashMethodAccess ma`'s qualifier or arguments flow to or from those of another `MDHashMethodAccess` call. */
```

luchua-bc replied:
I've simplified the implementation since it only checks calls in the same enclosing block and updated the comment.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+predicate hasAnotherHashCall(MethodAccess ma) {
+  exists(MDHashMethodAccess ma2, DataFlow::Node node1, DataFlow::Node node2 |
+    ma2 != ma and
+    node1.asExpr() = ma.getAChildExpr() and
```
Perhaps comment what you're expecting to see here? I'm a little surprised you're looking for non-qualifier child expressions flowing to or from the other call?

luchua-bc replied:
Simplified the implementation to only check calls in the same enclosing block with the same qualifier.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+/** Holds if `MethodAccess` is a single hashing call that is not invoked by a wrapper method. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not exists(MethodAccess ma2 | hasParentCall(ma2, ma)) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack to reduce FPs
```
```suggestion
  not hasParentCall(_, ma)) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack. This reduces FPs.
```
This seems like a very significant carve-out -- we're excluding any hash-method call with a caller? I also think considering `ma2` is completely free you probably don't need the recursion in `hasParentCall` and are simply excluding all methods that we know of any call to (i.e., only methods that are exclusively called by external code unseen by us will be counted, such as `main` and request handlers called by an external library?)

luchua-bc replied:
Updated

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+    or
+    exists(MethodAccess ma |
+      ma.getQualifier().(VarAccess).getVariable().getType() instanceof Interface and
+      ma.getAnArgument() = node.asExpr() // Method access of interface type variables requires runtime determination
```
Explain why we're sanitising all arguments to calls against an interface type?

luchua-bc replied:
Hi @smowton,

Thanks for reviewing the changes. For the current implementation with `MethodAccess`, I noticed it won't be able to detect concrete implementations of interface types at runtime, e.g. the following test case:
```
	// BAD - Invoke a wrapper implementation with a salt, which is not detected with an interface type variable.
	public String getSHA256Hash7(byte[] passphrase) throws NoSuchAlgorithmException, ClassNotFoundException, IllegalAccessException, InstantiationException {
		Class c = Class.forName("SHA512");
		HASH sha512 = (HASH) (c.newInstance());
		byte[] tmp = new byte[4];
		byte[] key = new byte[32 * 2];
		for (int i = 0; i < 2; i++) {
			sha512.init();
			tmp[3] = (byte) i;
			sha512.update(passphrase, 0, passphrase.length);
			System.arraycopy(sha512.digest(), 0, key, i * 32, 32);
		}
		return Base64.getEncoder().encodeToString(key);
	}
``` 

Without the exclusion of interface types, even there is an <code>arraycopy</code> method call to mix password digest with some other bytes, the query won't detect thus will report it. However, if I change the object type from the interface type <code>HASH</code> to the class type <code>SHA512</code>, the query can detect. That is, to change
```
		Class c = Class.forName("SHA512");
		HASH sha512 = (HASH) (c.newInstance());
```  
to
```
		SHA512 sha512 = new SHA512();
```

Is there a better approach to handle this scenario? Thanks a lot.

smowton replied:
OK, for the time being it's probably enough to note this in a comment and leave this for when the query is promoted out of experimental.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,153 @@
+    or
+    exists(AddExpr e | hasAddExpr(e, node.asExpr())) // password+salt
+    or
+    exists(ConditionalExpr ce | ce = node.asExpr()) // useSalt?password+":"+salt:password
```
From your comment -- perhaps you just want to sanitize conditional expressions with one arm itself passing the `isSanitizer` check?

luchua-bc replied:
For <code>ConditionalExpr</code>, both the conditional expression and individual arms work. I've modified the query to take one arm using <code>ce.getAChildExpr()</code> since it's more clear.

smowton reviewed (approved):
Happy with this, sending for final review

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,143 @@
+/** Holds if `MethodAccess` is a single hashing call that is not invoked by a wrapper method. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not hasParentCall(_, ma) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack. This reduces FPs.
```
This restriction is silly. It removes all results in methods that are being called from somewhere - that's removing a lot! And I don't see how that restriction is in any way related to the result being a FP or not. I think this restriction needs to be removed, and if that then produces too many FPs, then I think it's back to the drawing board. As-is I would have to reject the PR.

luchua-bc replied:
Hi @aschackmull,

Thanks for reviewing this PR. I believe this query is a valid one that handles an important scenario, and it is also doable especially with the help from both you and @smowton. 

In the new version I just committed, I restricted the sink to be the top of the call hierarchy that invokes a hashing method. It renders expected results for all the test cases and test projects I tested so I believe it's ready for review.

Please re-review. Thanks.

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql:
```diff
@@ -0,0 +1,188 @@
+/**
+ * @name Use of a hash function without a salt
+ * @description Hashed passwords without a salt are vulnerable to dictionary attacks.
+ * @kind path-problem
```
```suggestion
 * @kind path-problem
 * @problem.severity warning
 * @precision low
```

aschackmull reviewed (approved):
Added missing metadata.

luchua-bc commented:
Thanks @aschackmull for approving the request.
